text/0000-conditional-compilation-checking.md

```diff
@@ -0,0 +1,486 @@
+fn do_features() {}
+
+#[cfg(has_mumble_frotz)] // this is valid, because --cfg valid(...) was never used
+fn do_mumble_frotz() {}
+
+#[cfg(feature = "lasers")] // this is valid, because "lasers" is in the valid_values(feature) list
+fn shoot_lasers() {}
+
+#[cfg(feature = "monkeys")] // this is INVALID, because "monkeys" is not in
+                            // the valid_values(feature) list
+fn write_shakespear() {}
+```
+
+### Example: Checking both condition names and feature values
+
+```bash
+# This turns on checking for feature values and for condition names.
+rustc --cfg 'valid_values(feature, "zapping", "lasers")' \
+      --cfg 'valid(is_embedded, has_feathers)'
+      --cfg has_feathers \
+      --cfg 'feature="zapping"' \
+```
+
+
+```rust
+#[cfg(is_embedded)] // this is valid, and #[cfg] evaluates to disabled
+fn do_embedded() {}
+
+#[cfg(has_feathers)] // this is valid, and #[cfg] evalutes to enabled
+fn do_features() {}
+
+#[cfg(has_mumble_frotz)] // this is INVALID
+fn do_mumble_frotz() {}
+
+#[cfg(feature = "lasers")] // this is valid, because "lasers" is in the valid_values(feature) list
+fn shoot_lasers() {}
+
+#[cfg(feature = "monkeys")] // this is INVALID, because "monkeys" is not in
+                            // the valid_values(feature) list
+fn write_shakespear() {}
+```
+
+## Cargo support
+
+Cargo is ideally positioned to enable checking for `feature` flags, since Cargo knows the set of
+valid features. Cargo will invoke `rustc --cfg 'valid_values(feature, "...", ...)'`, so that
+checking for features is enabled. Optionally, Cargo could also specify the set of valid condition
+names.
+
+## Command line flags for checking conditional compilation
+
+To use this feature, you give `rustc` the set of condition names that are valid, by using the
+`--cfg valid(...)` flag. For example:
+
+```bash
+rustc --cfg 'valid(test, target_os, feature, foo, bar)' ...
+```
+
+> Note: This example shows `test` and `target_os`, but it is not necessary to specify these. All
+> condition names that are well-known to `rustc` are always permitted.
+
+For condition values, specify the set of values that are legal for a given condition name. For
+example:
+
+```bash
+ --cfg 'valid_values(feature, "foo", "bar", ..., "zot")'
+ ```
+
+## What do users need to do to use this?
+
+Most of the time, users will not invoke `rustc` directly. Instead, users run Cargo, which handles
+building command-lines for `rustc`. Cargo handles building the lists of valid condition names.
+
+> This reflects the final goal, which is turning on condition checking in Cargo. However, we will
+> need to do a gradual deployment:
+> 1. Users on `nightly` builds can run `cargo -Z check-cfg` to opt-in to checking.
+> 2. After stabilization, users can opt-in to conditional checking by using `cargo build --check-cfg`.
+> 3. Eventually, if the community experience is positive, Cargo could enable this check by default.

```

Comment:
I think cargo should _unconditionally_ provide the information to rustc, and then you can configure the lint with normal lints flags (allow, deny, `-D`).